File: AdventOfCode2022/day04.py
# -*- coding: utf-8 -*-
# Copyright 2022 by Alex Utter

import re
from aocd import get_data

def read_input(input):
    readline = lambda line: [int(x) for x in re.findall('[0-9]+', line)]
    return [readline(line) for line in input.splitlines()]

def part1(ranges):
    total = 0
    for [a0, a1, b0, b1] in ranges:
        a_in_b = (b0 <= a0 and a1 <= b1)
        b_in_a = (a0 <= b0 and b1 <= a1)
        if a_in_b or b_in_a: total += 1
    return total

def part2(ranges):
    total = 0
    for [a0, a1, b0, b1] in ranges:
        ovr1 = (a0 <= b0 and b0 <= a1)
        ovr2 = (a0 <= b1 and b1 <= a1)
        ovr3 = (b0 <= a0 and a0 <= b1)
        ovr4 = (b0 <= a1 and a1 <= b1)
        if ovr1 or ovr2 or ovr3 or ovr4: total += 1
    return total

TEST= \
'''
2-4,6-8
2-3,4-5
5-7,7-9
2-8,3-7
6-6,4-6
2-6,4-8
'''

if __name__ == '__main__':
    test = read_input(TEST.strip())
    input = read_input(get_data(day=4, year=2022))
    assert(part1(test) == 2)
    assert(part2(test) == 4)
    print(f'Part 1: {part1(input)}')
    print(f'Part 2: {part2(input)}')

File: AdventOfCode2016/src/bin/day04.rs
/// Advent of Code 2016, Day 4
/// Copyright 2023 by Alex Utter

use aocfetch;

use std::collections::BinaryHeap;
use std::collections::HashMap;

struct Data {
    name: String,
    id: u32,
    hash: String,
}

impl Data {
    // Parse data entry from a line of text.
    fn new(line: &str) -> Self {
        let raw: Vec<&str> = line.trim().split('-').collect();
        let name = raw[0..raw.len()-1].join("-");
        let last = raw.last().unwrap();
        let len = last.len();
        let hash = last[len-6..len-1].to_string();
        let id:u32 = last[0..len-7].parse().unwrap();
        Data {name:name, id:id, hash:hash}
    }

    // Decrypt the label string.
    fn decrypt(&self) -> String {
        const BASE: u32 = 'a' as u32;
        let mut result = String::new();
        for ch in self.name.chars() {
            if ch == '-' {
                result.push(' ');
            } else {
                let idx = (self.id + ch as u32 - BASE) % 26;
                result.push(char::from_u32(BASE + idx).unwrap());
            }
        }
        return result;
    }

    // Does the hash match the expected value?
    fn valid(&self) -> bool {
        // Count occurence of each letter.
        let mut counts: HashMap<char,u64> = HashMap::new();
        for ch in self.name.chars() {
            if ch == '-' {continue;}
            *counts.entry(ch).or_insert(0) += 1;
        }

        // Sort the list of letter counts.
        let mut heap = BinaryHeap::new();
        for (ch, ct) in counts {
            let ch_sort = (char::MAX as u32) - (ch as u32);
            heap.push((ct,ch_sort,ch));
        }
        // Expected hash is the top five characters.
        let (_,_,h1) = heap.pop().unwrap();
        let (_,_,h2) = heap.pop().unwrap();
        let (_,_,h3) = heap.pop().unwrap();
        let (_,_,h4) = heap.pop().unwrap();
        let (_,_,h5) = heap.pop().unwrap();
        let hash = format!("{}{}{}{}{}", h1, h2, h3, h4, h5);
        return hash == self.hash;
    }
}

fn part1(input: &str) -> i64
{
    let mut total = 0i64;
    for line in input.trim().lines() {
        let data = Data::new(line);
        if data.valid() {total += data.id as i64;}
    }
    return total;
}

fn part2(input: &str) -> i64
{
    for line in input.trim().lines() {
        let data = Data::new(line);
        if data.valid() && data.decrypt() == "northpole object storage" {
            return data.id as i64;
        }
    }
    0
}

const TEST: &str = "\
aaaaa-bbb-z-y-x-123[abxyz]
a-b-c-d-e-f-g-h-987[abcde]
not-a-real-room-404[oarel]
totally-real-room-200[decoy]";

fn main() {
    // Fetch input from server.
    let input = aocfetch::get_data(2016, 4).unwrap();

    // Unit tests on provided examples
    assert_eq!(part1(TEST), 1514);

    // Solve for real input.
    println!("Part 1: {}", part1(&input));
    println!("Part 2: {}", part2(&input));
}
